src/wp-includes/html-api/class-wp-html-doctype-info.php

```diff
@@ -0,0 +1,269 @@
+<?php
+/**
+ * HTML API: WP_HTML_Doctype_Info class
+ *
+ * @package WordPress
+ * @subpackage HTML-API
+ * @since 6.7.0
+ */
+
+/**
+ * Core class used by the HTML API processor to represent a DOCTYPE declaration.
+ *
+ * @since 6.7.0
+ */
+class WP_HTML_Doctype_Info {
+	/**
+	 * The name of the DOCTYPE.
+	 *
+	 * @var string|null
+	 */
+	private $name;
+
+	/**
+	 * The public identifier of the DOCTYPE.
+	 *
+	 * @var string|null
+	 */
+	private $public_identifier;
+
+	/**
+	 * The system identifier of the DOCTYPE.
+	 *
+	 * @var string|null
+	 */
+	private $system_identifier;
+
+	/**
+	 * Whether the DOCTYPE token force-quirks flag is set.
+	 *
+	 * @var bool
+	 */
+	private $force_quirks_flag;
+
+	/**
+	 * Constructor.
+	 *
+	 * The arguments to this constructor correspond to the "DOCTYPE token" as defined in the
+	 * HTML specification.
+	 *
+	 * > DOCTYPE tokens have a name, a public identifier, a system identifier,
+	 * > and a force-quirks flag.
+	 *
+	 * @see https://html.spec.whatwg.org/multipage/parsing.html#tokenization
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param string|null $name              The name of the DOCTYPE.
+	 * @param string|null $public_identifier The public identifier of the DOCTYPE.
+	 * @param string|null $system_identifier The system identifier of the DOCTYPE.
+	 * @param bool        $force_quirks_flag Whether the DOCTYPE token force-quirks flag is set.
+	 */
+	public function __construct(
+		?string $name,
+		?string $public_identifier,
+		?string $system_identifier,
+		bool $force_quirks_flag
+	) {
+		$this->name              = $name;
+		$this->public_identifier = $public_identifier;
+		$this->system_identifier = $system_identifier;
+		$this->force_quirks_flag = $force_quirks_flag;
+	}
+
+	/**
+	 * Gets the name of the DOCTYPE.
+	 *
+	 * @since 6.7.0
+	 *
+	 * @return string The name of the DOCTYPE.
+	 */
+	public function get_name(): string {
+		return $this->name ?? '';
+	}
```

Comment:
given that this basically serves as a record class with an intelligent constructor, do you see much value in hiding these behind `private`?

what if the properties were all `public` and set upon instantiation, including empty strings when not set?

Reply:
I don't think it's OK to set these properties, that's one reason to use getters.

Reply:
There's also the `null/""` difference in some of the quirks mode rules, but that's less relevant since I extracted the quirks mode algorithm to another method.

Reply:
it's not okay to set them, but also seems needless to recreate state-oriented interfaces that do nothing. I know "once public always public" occurs here, but I also see these as being closed to extension or updates.

we can also document them saying "don't change these" but I don't know why someone _would_ when the class doesn't do anything else. it doesn't support serialization and it shouldn't be re-parsing itself to change the suggested document mode.

in essence, they are read only unless someone changes them and re-reads them, and I'm not sure what the point of that would be.